File: Trie.java
class Trie{
    Trie child[];
    boolean isEnd;
    Trie(){
        child = new Trie[26];
        isEnd = false;
    }
    public void insert(String word){
        Trie curr = this;
        for(char c : word.toCharArray()){
            int idx = c-'a';
            if(curr.child[idx] == null){
                curr.child[idx] = new Trie();
            }
            curr = curr.child[idx];
        }
        curr.isEnd = true;
    }
    public boolean search(String word){
        Trie curr = this;
        for(char c : word.toCharArray()){
            int idx = c-'a';
            if(curr.child[idx] == null){
                return false;
            }
            curr = curr.child[idx];
        }
        return curr.isEnd;
    }
    public boolean startsWith(String prefix){
        Trie curr = this;
        for(char c : prefix.toCharArray()){
            int idx = c-'a';
            if(curr.child[idx] == null){
                return false;
            }
            curr = curr.child[idx];
        }
        return true;
    }
    
    public static void main(String[] args) {
        Trie root = new Trie();

        String [] sen = {"my", "name", "is", "kanai"};
        for(String s: sen){
            root.insert(s);
        }

        System.out.println(root.search("kanai"));
        System.out.println(root.search("kan"));
        System.out.println(root.startsWith("naruto"));

    }
}

File: trie.py
class TrieNode:
    def __init__(self):
        self.children = {}
        self.is_end_of_word = False

class Trie:
    def __init__(self):
        self.root = TrieNode()

    def insert(self, word):
        current = self.root
        for char in word:
            if char not in current.children:
                current.children[char] = TrieNode()
            current = current.children[char]
        current.is_end_of_word = True

    def search(self, word):
        current = self.root
        for char in word:
            if char not in current.children:
                return False
            current = current.children[char]
        return current.is_end_of_word

    def starts_with(self, prefix):
        current = self.root
        for char in prefix:
            if char not in current.children:
                return False
            current = current.children[char]
        return True

# Example usage:
trie = Trie()

# Inserting words
trie.insert("apple")
trie.insert("banana")
trie.insert("orange")

# Searching words
print(trie.search("apple"))  # True
print(trie.search("banana"))  # True
print(trie.search("orange"))  # True
print(trie.search("grape"))  # False

# Searching prefixes
print(trie.starts_with("app"))  # True
print(trie.starts_with("ban"))  # True
print(trie.starts_with("ora"))  # True
print(trie.starts_with("gra"))  # False
